pmlc/dialect/pxa/transforms/dealloc_placement.cc

```diff
@@ -57,102 +56,48 @@ struct DeallocPlacementPass
-      for (unsigned i = 0; i < args.size(); ++i) {
-        placeDealloc(cast<Value>(args[i]), &forOp.getBody()->front(),
-                     forOp.getBody(), i, onPack);
-      }
-    });
-  }
 
-  // This function dealloc ref if possible, which is the allocated memory
-  // reference. firstOp is generally the allocation operaion. For scf.for
-  // arguments, it is virtually the first operation in the loop. argNumber is
-  // scf.for argument order number, which is useless for normal deallocation.
-  template <typename Callback>
-  void placeDealloc(Value ref, Operation *firstOp, Block *allocBlock,
-                    int argNumber, Callback onPack) {
-    Operation *lastOp = firstOp;
-    OpOperand *lastUse = nullptr;
-    for (auto &itUse : getIndirectUses(ref)) {
-      auto use = itUse.getOwner();
-      IVLOG(3, "  use: " << debugString(*use));
+      Operation *lastOp = alloc;
+      OpOperand *lastUse = nullptr;
+      Block *allocBlock = alloc.getOperation()->getBlock();
+      for (auto &itUse : getIndirectUses(alloc)) {
+        auto use = itUse.getOwner();
+        IVLOG(3, "  use: " << debugString(*use));
+
+        auto ancestor = allocBlock->findAncestorOpInBlock(*use);
+        assert(ancestor && "use and alloc do not have a common ancestor");
+        IVLOG(3, "  ancestor: " << debugString(*use));
 
-      auto ancestor = allocBlock->findAncestorOpInBlock(*use);
-      assert(ancestor && "use and alloc do not have a common ancestor");
-      IVLOG(3, "  ancestor: " << debugString(*use));
+        if (isa<ReturnOp>(ancestor)) {
+          IVLOG(3, "  return");
+          return;
+        }
 
-      if (isa<ReturnOp>(ancestor)) {
-        IVLOG(3, "  return");
+        if (!ancestor->isBeforeInBlock(lastOp)) {
+          lastOp = ancestor;
+          lastUse = &itUse;
+        }
+      }
+      IVLOG(3, "  last ancestor: " << debugString(*lastOp));
+      if (auto packOp = dyn_cast<stdx::PackOp>(lastOp)) {
+        IVLOG(3, "  pack op");
+        // Alloc 'escapes' via a pack, call our callback to handle
+        onPack(lastUse->getOperandNumber());
         return;
       }
 
-      if (!ancestor->isBeforeInBlock(lastOp)) {
-        lastOp = ancestor;
-        lastUse = &itUse;
+      Operation *nextOp = lastOp->getNextNode();
+      if (!nextOp) {
+        IVLOG(3, "  terminator");
+        return;
       }
-    }
-    IVLOG(3, "  last ancestor: " << debugString(*lastOp));
-    if (auto packOp = dyn_cast<stdx::PackOp>(lastOp)) {
-      IVLOG(3, "  pack op");
-      // Alloc 'escapes' via a pack, call our callback to handle
-      onPack(lastUse->getOperandNumber());
-      return;
-    }
 
-    Operation *nextOp = lastOp->getNextNode();
-    if (!nextOp) {
-      IVLOG(3, "  terminator");
-      return;
-    }
-
-    IVLOG(3, "  next operation: " << debugString(*nextOp));
-    OpBuilder builder(nextOp);
-    builder.create<DeallocOp>(firstOp->getLoc(), lastUse->get());
-    if (argNumber >= 0) {
```

Comment:
after this `if(argNumber >= 0) `code, I believe that yao are trying to follow the tile_pxa pass style,  it will alloc a new memref to replace the function argument which is used to initial scf.for loop carriable variable(iterator), but don't like tile_pxa, he don't eliminate this new alloc memref. I don't think this action is necessary.